text/3493-precise-pre-release-cargo-update.md

```diff
@@ -0,0 +1,139 @@
+
+If a user does want to select a pre-release version they are able to do so by explicitly requesting Cargo to update to that version.
+This is done by passing the `--precise` flag to Cargo.
+Cargo will refuse to select pre-release versions that are "incompatible" with the requirement in the projects `Cargo.toml`.
+A pre-release version is considered compatible for a precise upgrade if its major, minor, and patch versions are compatible with the requirement, ignoring its pre-release version.
+`x.y.z-pre0` is considered compatible with `a.b.c` when requested `--precise`ly if `x.y.z` is semver compatible with `a.b.c` and `a.b.c` `!=` `x.y.z`.
+
+Consider a `Cargo.toml` with this `[dependencies]` section
+
+```
+[dependencies]
+example = "1.0.0"
+```
+
+It is possible to update to `1.2.0-pre0` because `1.2.0` is semver compatible with `1.0.0`
+
+```
+> cargo update -p example --precise 1.2.0-pre0
+    Updating crates.io index
+    Updating example v1.0.0 -> v1.2.0-pre0
+```
+
+It is not possible to update to `2.0.0-pre0` because `2.0.0` is not semver compatible with `1.0.0`
+
+```
+> cargo update -p example --precise 2.0.0-pre0
+    Updating crates.io index
+error: failed to select a version for the requirement `example = "^1"`
+candidate versions found which didn't match: 2.0.0-pre0
+location searched: crates.io index
+required by package `tmp-oyyzsf v0.1.0 (/home/ethan/.cache/cargo-temp/tmp-OYyZsF)`
+```
+
+Once `1.2.0` is released `cargo update` will warn the user that a stable version is available
+
+```
+> cargo update
+warning: version `1.2.0` of `example` has been released but `1.2.0-pre0` was previously selected
+note: to use the stable version run `cargo update -p example --precise 1.2.0`
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Consider this table where `a.b.c` is compatible with `x.y.z` and `x.y.z > a.b.c`
+
+| Cargo.toml spec | Cargo.lock version | Target version | Selected by cargo update  | Selected by cargo update --precise  |
+| --------------- | ------------------ | -------------- | ------------------------- | ----------------------------------- |
+| `a.b.c`         | `a.b.c`            | `x.y.z`        | ✅                        | ✅                                  |
+| `a.b.c`         | `a.b.c`            | `x.y.z-pre0`   | ❌                        | ✅                                  |
+| `a.b.c`         | `x.y.z-pre0`       | `x.y.z-pre1`   | ❌                        | ✅                                  |
+| `a.b.c-pre0`    | `a.b.c-pre0`       | `a.b.c-pre1`   | ✅¹                       | ✅                                  |
+| `a.b.c`         | `a.b.c`            | `a.b.c-pre0`   | ❌                        | ❌                                  |
+| `a.b.c-pre0`    | `a.b.c-pre0`       | `x.y.z`        | ❌²                       | ✅                                  |
```

Comment:
Why does this not keep the existing "✅ | ✅" behavior like the other back-compat line above?

Reply:
I did not know that way the current behavior. It's a shame as `a.b.c-pre0` is likely incompatible with `x.y.z`. I'll update the RFC to keep it backwards compatible.

Reply:
https://github.com/rust-lang/rfcs/pull/3493/commits/a3696be465ef07e5a71326ef91f362dc84da0c7c